File: java/Arrays/Day26/ReversePairs.java
package Day26;
import java.util.*;

public class ReversePairs {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter number of elements: ");
        int n = sc.nextInt();

        System.out.print("Enter " + n + " " + "numbers: ");
        int[] arr = new int[n];
        for (int i = 0; i < n; i++) {
            arr[i] = sc.nextInt();
        }



        // -------- BRUTE APPROACH FOR REVERSE PAIRS ----------
        // int cnt = 0;
        // for (int i = 0; i < n - 1; i++) {
        //     for (int j = i + 1; j < n; j++) {
        //         if (arr[i] > 2 * arr[j]) {
        //             cnt += 1;
        //         }
        //     }
        // }

        // System.out.print("Reverse Pairs: " + cnt);





        // -------- OPTIMAL APPROACH -----------
        System.out.print("Reverse Pairs: " + mergeSort(arr, 0, n - 1));

        sc.close();
    }



    // -------- OPTIMAL APPROACH -----------
    private static void merge(int[] arr, int low, int mid, int high) {
        List<Integer> temp = new ArrayList<>();
        int left = low;
        int right = mid + 1;

        while (left <= mid && right <= high) {
            if (arr[left] <= arr[right]) {
                temp.add(arr[left]);
                left++;
            }
            else {
                temp.add(arr[right]);
                right++;
            }
        }

        while (left <= mid) {
            temp.add(arr[left]);
            left++;
        }

        while (right <= high) {
            temp.add(arr[right]);
            right++;
        }

        for (int i = low; i <= high; i++) {
            arr[i] = temp.get(i - low);
        }
    }




    private static int countPairs(int[] arr, int low, int mid, int high) {
        int right = mid + 1;
        int cnt = 0;

        for (int i = low; i <= mid; i++) {
            while (right <= high && (long) arr[i] > 2L * arr[right]) right++;
            cnt += (right - (mid + 1));
        }

        return cnt;
    }





    private static int mergeSort(int[] arr, int low, int high) {
        int cnt = 0;
        if (low >= high) return cnt;

        int mid = (low + high) / 2;
        cnt += mergeSort(arr, low, mid);
        cnt += mergeSort(arr, mid + 1, high);
        cnt += countPairs(arr, low, mid, high);
        merge(arr, low, mid, high);

        return cnt;
    }
}

File: python/Arrays/day26/reversePairs.py
# -------- OPTIMAL APPROACH --------
def merge_sort(arr, low, high):
    cnt = 0
    if low >= high:
        return cnt
    
    mid = (low + high) // 2
    cnt += merge_sort(arr, low, mid)
    cnt += merge_sort(arr, mid + 1, high)
    cnt += count_pairs(arr, low, mid, high)
    merge(arr, low, mid, high)

    return cnt



def merge(arr, low, mid, high):
    temp = []
    left = low
    right = mid + 1

    while left <= mid and right <= high:
        if arr[left] <= arr[right]:
            temp.append(arr[left])
            left += 1
        
        else:
            temp.append(arr[right])
            right += 1
    
    while left <= mid:
        temp.append(arr[left])
        left += 1
    
    while right <= high:
        temp.append(arr[right])
        right += 1

    for i in range(low, high + 1):
        arr[i] = temp[i - low]




def count_pairs(arr, low, mid, high):
    right = mid + 1
    cnt = 0
    for i in range(low, mid + 1):
        while right <= high and arr[i] > 2 * arr[right]:
            right += 1
        
        cnt += (right - (mid + 1))
    
    return cnt






# -------- MAIN --------
n = int(input("Enter number of element: "))

arr = list(map(int, input(f"Enter {n} numbers: ").split()))


# -------- OPTIMAL APPROACH --------
print("Reverse Pairs:", merge_sort(arr, 0, n - 1))



# -------- BRUTE FORCE APPROACH ---------
# cnt = 0
# for i in range(n - 1):
#     for j in range(i + 1, n):
#         if arr[i] > 2 * arr[j]:
#             cnt += 1
    
# print("Reverse Pairs:", cnt)